pandas/core/index.py

```diff
@@ -391,6 +395,11 @@ def _shallow_copy(self, values=None, infer=False, **kwargs):
 
         if infer:
             attributes['copy'] = False
+            if self._infer_as_myclass:
```

Comment:
I understand. The point is that maybe if we have a class attribute, we don't need the instance one (e.g. passing ``infer=True`` at all). OR (if you think we need both) I would name them the same, and only use the instance passed one if its not ``None``